server/src/main/java/org/elasticsearch/gateway/PersistedClusterStateService.java

```diff
@@ -547,6 +523,10 @@ public void close() throws IOException {
         boolean fullStateWritten = false;
         private final AtomicBoolean closed = new AtomicBoolean();
 
+        // The size of the document buffer that was used for the last write operation, used as a hint for allocating the buffer for the
+        // next one.
+        private int documentBufferUsed;
```

Comment:
I debated keeping the whole buffer around for the next time we need to write the cluster state, maybe shrinking it based on some heuristics (e.g. it was <50% used this time), but decided against it since we don't write the cluster state out _that_ much so it might just sit there consuming 10s of MBs of heap (or more).